### PR TITLE
Add test for stateful duplicate issue

### PR DESCRIPTION
**Pull request by Warchant**

### CODE REVIEW

File: include/veriblock/pop/blockchain/base_block_tree.hpp [hunk 1]
```diff
@@ -849,21 +850,22 @@ struct BaseBlockTree {
...
                   preserveBlocksBehindFinal);
 
     index_t* finalizedBlock = &index;
 
-    // prereq is not met - finalized block must be on active chain
-    if (!activeChain_.contains(finalizedBlock)) {
-      return state.Invalid("block-not-on-active-chain");
+    if (finalizedBlock == getRoot()) {
+      return true;
     }
```
Comment (blackyblack) [inline include/veriblock/pop/blockchain/base_block_tree.hpp:863]: Why this check?

File: src/pop/blockchain/alt_block_tree.cpp [hunk 2]
```diff
@@ -568,6 +571,27 @@ bool AltBlockTree::setState(index_t& to, ValidationState& state) {
   return success;
 }
 
+void AltBlockTree::doFinalize() {
+  ValidationState state;
+  auto* tip = getBestChain().tip();
+  VBK_ASSERT(tip && "must be bootstrapped");
```
Comment (blackyblack) [inline src/pop/blockchain/alt_block_tree.cpp:577]: It seems you prefer to assert != nullptr. Make it consistent over new code:
```
VBK_ASSERT(tip != nullptr && "must be bootstrapped");
```

File: src/pop/blockchain/alt_block_tree.cpp [hunk 3]
```diff
@@ -568,6 +571,27 @@ bool AltBlockTree::setState(index_t& to, ValidationState& state) {
...
+
+  int32_t maxReorg = (int32_t)getParams().getMaxReorgDistance();
+  if (maxReorg > getRoot().getHeight()) {
+    return;
+  }
```
Comment (blackyblack) [inline src/pop/blockchain/alt_block_tree.cpp:582]: It looks like finalization won't work until root reaches maxReorg height, eg if maxReorg is 100000 finalization won't work until root is on height 100000.

File: src/pop/blockchain/alt_block_tree.cpp [hunk 4]
```diff
@@ -698,6 +724,8 @@ bool AltBlockTree::loadTip(const AltBlockTree::hash_t& hash,
     tip = tip->pprev;
   }
 
+  // finalize a block
+
```
Comment (blackyblack) [inline src/pop/blockchain/alt_block_tree.cpp:728]: What is this?

File: src/pop/blockchain/alt_block_tree.cpp [hunk 5]
```diff
@@ -745,9 +773,11 @@ bool AltBlockTree::finalizeBlockImpl(index_t& index,
...
   auto* finalizedIndex = vbk().getBestChain()[firstBlockHeight];
   VBK_ASSERT_MSG(finalizedIndex != nullptr, "Invalid VBK tree state");
-  if (!vbk().finalizeBlock(*finalizedIndex, state)) {
-    return state.Invalid("vbktree-finalize-error");
-  }
+  bool result = vbk().finalizeBlock(*finalizedIndex, state);
+  VBK_ASSERT_MSG(result,
+                 "Failed to finalize block %s: %s",
+                 finalizedIndex->toShortPrettyString(),
+                 state.toString());
```
Comment (blackyblack) [inline src/pop/blockchain/alt_block_tree.cpp:780]: It seems that finalizeBlock should never fail. Put this assert inside finalizeBlock() and do not return anything from it.

**Review by blackyblack — CHANGES_REQUESTED** [review]

File: include/veriblock/pop/blockchain/base_block_tree.hpp (hunk 1 above)
Comment (Warchant) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:863]: It's an early return

Comment (blackyblack) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:863]: Yes but it is covered but the code below. Does not look like it makes any speed up.

File: src/pop/blockchain/alt_block_tree.cpp (hunk 3 above)
Comment (Warchant) [reply src/pop/blockchain/alt_block_tree.cpp:582]: you're right, it's a typo. we should compare maxReorg with tip height, not root height

File: include/veriblock/pop/blockchain/base_block_tree.hpp (hunk 1 above)
Comment (Warchant) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:863]: this case will be triggered often, and instead of iterating over `tips` and doing some other looping we could just early return, this is faster by definition (we don't invoke code we don't need)

File: include/veriblock/pop/blockchain/base_block_tree.hpp [hunk 6]
```diff
@@ -951,7 +952,7 @@ struct BaseBlockTree {
       ptr = ptr->pprev;
     }
 
-    return true;
+    return;
```
Comment (blackyblack) [inline include/veriblock/pop/blockchain/base_block_tree.hpp:955]: Useless.

File: include/veriblock/pop/blockchain/blocktree.hpp [hunk 7]
```diff
@@ -211,15 +211,14 @@ struct BlockTree : public BaseBlockTree<Block> {
...
+  void finalizeBlock(index_t& index) {
     (void)index;
-    (void)state;
     // TODO: currently can not procced finalization for the vbk and btc trees,
     // need more investigation
 
     // return this->finalizeBlockImpl(
     //     index, param_->preserveBlocksBehindFinal(), state);
-    return true;
+    return;
```
Comment (blackyblack) [inline include/veriblock/pop/blockchain/blocktree.hpp:221]: Remove.

File: src/pop/blockchain/pop/vbk_block_tree.cpp [hunk 8]
```diff
@@ -434,10 +433,8 @@ bool VbkBlockTree::finalizeBlockImpl(index_t& index,
   firstBlockHeight = std::max(bootstrapBlockHeight, firstBlockHeight);
   auto* finalizedIndex = btc().getBestChain()[firstBlockHeight];
   VBK_ASSERT_MSG(finalizedIndex != nullptr, "Invalid BTC tree state");
-  if (!btc().finalizeBlock(*finalizedIndex, state)) {
-    return state.Invalid("btctree-finalize-error");
-  }
-  return base::finalizeBlockImpl(index, preserveBlocksBehindFinal, state);
+  btc().finalizeBlock(*finalizedIndex);
+  return base::finalizeBlockImpl(index, preserveBlocksBehindFinal);
```
Comment (blackyblack) [inline src/pop/blockchain/pop/vbk_block_tree.cpp:437]: Remove `return`:
```
btc().finalizeBlock(*finalizedIndex);
base::finalizeBlockImpl(index, preserveBlocksBehindFinal);
```

File: test/pop/blockchain/altblock_finalization_test_3.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    altparam.mEndorsementSettlementInterval = 50;
+    altparam.mPreserveBlocksBehindFinal = 50;
+  }
+
+  std::vector<AltBlock> chain = {altparam.getBootstrapBlock()};
+};
+
+TEST_F(StatefulDuplicateWhenFinalized, VBK) {
+  alttree.onInvalidBlockConnected.connect(
+      [](BlockIndex<AltBlock> &/*index*/, ValidationState &state) {
```
Comment (blackyblack) [inline test/pop/blockchain/altblock_finalization_test_3.cpp:28]: Rename `state` - it hides PopTestFixture::state field.

**Review by blackyblack — CHANGES_REQUESTED** [review]
